lib/logreader.c

```diff
@@ -402,6 +423,8 @@ log_reader_work_finished(void *s, gpointer arg)
       log_proto_server_reset_error(self->proto);
       log_reader_update_watches(self);
     }
+
+  log_reader_set_work_in_progress(self, FALSE);
```

Comment:
my understanding is that even `log_reader_work_perform` and `log_reader_work_finished` can be called from another thread, but no multiple workers of the same reader can be duplicated, so even if the reader is threaded tha same sequence will be used that can see in `log_reader_io_handle_in`


log_pipe_ref
log_reader_work_perform
log_reader_work_finished
log_pipe_unref


but you made me rethink this once more, and will now move the work in progress TRUE setter to the location I originally planned, to the `log_reader_work_perform` function instead to keep the flag really synced with the work in progress state (NOTE: it would be nice to use MainLoopIOWorkerJob working member instead, but i was not sure what else i can screw up with that, so added a dedicated member instead that wil surely behave the same way both for the threaded and the none-threaded version)